readme.md
# Football Simulation Engine
---
## Overview
This module was designed to allow the simulation of football (soccer) matches between two teams. This module allows for an iterative football match to be played out given initial "state" of players. 

The module consists of three functions that:
 - Initiate a match
 - complete an iteration / movement
 - switch team sides / start second half

For examples of how this module can be used, see:
* [A narrated video of a match.](https://youtu.be/yxTXFrAZCdY)
* [An example Node implementation of a Football Simulator with a GUI](https://github.com/GallagherAiden/footballsimulationexample) Note: not test on latest version
* [An example Node implementation for the 2018 World Cup](https://github.com/GallagherAiden/worldcup2018simulator) Note: not test on latest version
---
## Latest Version (4.0.0)
- fixed: red cards didn't send players off the pitch correctly
- fixed: player skills incorrectly assigned for tackles and slide tackles
- fixed: corners and goal kicks not correctly assigned / calculated
- fixed: intentPOS sometimes returned as null
- fixed: second half returned kickoffTeam with the ball, now secondTeam
- new: added ability to set the width of the goal to make the game more customisable
- [Full and Past changelogs are available here.](history.md)

---
## Install
Make sure your version of Node.js is at least 7.6.0. (The 'async/await' function is used)

```npm install --save footballsimulationengine```

---
## Initiate Game
This function is a promise that expects two teams and pitch information in JSON format (JSON format is described below). The function will return match details including player start positions, match statistics, ball position and an iteration log.

#### Example Call
```
initiateGame(team1, team2, pitch).then(function(matchDetails){
  console.log(matchDetails);
}).catch(function(error){
  console.error("Error: ", error);
})
```

## Play Iteration
This function is a promise that expects a single JSON input. This input will match the output of the initiate game function and will make the next move of both teams, resolving the final position of each player.

#### Example Call
```
playIteration(matchDetails).then(function (matchDetails) {
  console.log(matchDetails);
}).catch(function(error){
  console.error("Error: ", error);
}
```
## Start Second Half (Switch Sides)
This function is a promise that switches the side of each team, as happens typically at the end of a half. This uses the output from either an initiate game or a play iteration.

#### Example Call
```
startSecondHalf(matchDetails).then(function (matchDetails) {
  console.log(matchDetails);
}).catch(function(error){
  console.error("Error: ", error);
}
```
---
## Recommendations
* Users can determine how many iterations make a half
* Test data found in `init_config` is the data used for testing
* Iteration logs give an overview of what has happened in the iteration
* Uncomment console.log(output) to receive iteration by iteration information of each players iteration action, movement, original position and final position (start POS).

## Additional Information
* Get in touch with any questions [email](mailto:aiden.g@live.co.uk)
* See a match example [here](https://youtu.be/yxTXFrAZCdY)
* Raise issues in [GitHub](https://github.com/GallagherAiden/footballSimulationEngine/issues)
---
# Examples:
Examples are baked into the file system (>v2.1.0) in the `init_config` directory:
 - `index.js` : example function usages
 - `team1.json` : example json for a primary team
 - `team2.json` : example json for a secondary team
 - `pitch.json` : example json for pitch details
 - `iteration.json` : shows what the overall output given for each iteration

#### Example Team JSON
Each team must have the following information and contain 11 players.
* A start position for each player, with both teams given start positions as if they were the team at the top of a vertical pitch (shooting to the bottom of the screen). The currentPOS object should be [widthPosition, heightPosition] where the height placement should not be a greater number than half the pitch height.
* skills should not exceed 100
* skill.jumping refers to height a player can jump in centimetres (so can and probably should be greater than 100).
```
{
  "name": "Team1",
  "players": [{
      "name": "Player",
      "position": "GK",
      "rating": "99",
      "skill": {
        "passing": "99",
        "shooting": "99",
        "tackling": "99",
        "saving": "99",
        "agility": "99",
        "strength": "99",
        "penalty_taking": "99",
        "jumping": "300"
      },
      "currentPOS": [60,0],
      "fitness": 100,
      "injured": false
    }...],
  "manager": "Aiden"
}
```

#### Example Pitch JSON
Pitch has been tested for width of 120 - 680 and height of 600 - 1050 and a goal width of 90. The below is the current provided pitch size.
```
{
	"pitchWidth": 680,
	"pitchHeight": 1050,
  goalWidth: 90
}
```

#### Example Match Details

history.md
# History of footballSimulationEngine

## Version 4.0.0
- fixed: red cards didn't send players off the pitch correctly
- fixed: player skills incorrectly assigned for tackles and slide tackles
- fixed: corners and goal kicks not correctly assigned / calculated
- fixed: intentPOS sometimes returned as null
- fixed: second half returned kickoffTeam with the ball, now secondTeam
- new: added ability to set the width of the goal to make the game more customisable

## Version 3.0.1
- added new test scripts
- code fixes for assigning actions
- code fixes for determining offside
- code fixes for setting positions 
- code fixes for when ball moving

## Version 3.0.0
- altered 'startPOS' to 'currentPOS' to better reflect what the variable is used for
- altered 'relativePOS' to 'intentPOS' to better reflect what what the variable is used for
- added 100+ test cases
- cleaned 'setPositions.js' and created/cleaned 'setFreekicks.js' to improve readability and reduce duplication
- improvement to corners by putting players in the box
- players with red cards are removed from the pitch
- penalty taking improvements
- skills used more intensively i.e. shooting can be both on/off target determined by skill
- add player specific stats including tackles, passes, shots, saves and goals
---
## Version 2.2.0
- fix closestPlayer Report
- enhance test cases
- keep ball with player when they run or sprint
- improve run/sprint direction
- improve passing
- improve Error reporting
- track shot over time
- add saving
- add simple player marking
- add test simulation data
---
## Version 2.1.2
- security fix for https://nvd.nist.gov/vuln/detail/CVE-2018-16487
- limit ability for players to become 'stuck'
- checks to reduce occurances of players.hasBall = true when the ball is not with the player
---
## Version 2.1.0
- Added to allow more tackles and fouls
- Added fitness measure (currently has no affect on the game - see later versions)
- Remove promises from the internal functions (still required for three main functions as above)
- Includes a test suite (still in development)
- Improvements to ball movement to give movement over time instead of a 'jump'
- refactoring to improve source code readability
- Removed dependency on async function
- Resolved error where players leave pitch
- Provided setup data - e.g. a pitch JSON and two team JSONs for easier setup
- Allow interception of the ball whilst in transit
- Included ability to set a players action for each iteration to enforce a specific move

lib/actions.js
const common = require('../lib/common')
const setPositions = require('../lib/setPositions')

function selectAction(possibleActions) {
  let goodActions = []
  for (const thisAction of possibleActions) {
    let tempArray = Array(thisAction.points).fill(thisAction.name)
    goodActions = goodActions.concat(tempArray)
  }
  if (goodActions[0] == null) return 'wait'
  return goodActions[common.getRandomNumber(0, goodActions.length - 1)]
}

function findPossActions(player, team, opposition, ballX, ballY, matchDetails) {
  let possibleActions = populateActionsJSON()
  const [, pitchHeight] = matchDetails.pitchSize
  let params = []
  let {
    hasBall, originPOS
  } = player
  if (hasBall === false) params = playerDoesNotHaveBall(player, ballX, ballY, matchDetails)
  else if (originPOS[1] > (pitchHeight / 2)) params = bottomTeamPlayerHasBall(matchDetails, player, team, opposition)
  else params = topTeamPlayerHasBall(matchDetails, player, team, opposition)
  return populatePossibleActions(possibleActions, ...params)
}

function topTeamPlayerHasBall(matchDetails, player, team, opposition) {
  let playerInformation = setPositions.closestPlayerToPosition(player, opposition, player.currentPOS)
  const [pitchWidth, pitchHeight] = matchDetails.pitchSize
  let {
    position, currentPOS, skill
  } = player
  if (position === 'GK' && oppositionNearPlayer(playerInformation, 10, 25)) return [0, 0, 10, 0, 0, 0, 0, 10, 0, 40, 40]
  else if (position === 'GK') return [0, 0, 50, 0, 0, 0, 0, 10, 0, 20, 20]
  else if (onBottomCornerBoundary(currentPOS, pitchWidth, pitchHeight)) return [0, 0, 20, 80, 0, 0, 0, 0, 0, 0, 0]
  else if (checkPositionInBottomPenaltyBox(currentPOS, pitchWidth, pitchHeight)) {
    return topTeamPlayerHasBallInBottomPenaltyBox(matchDetails, player, team, opposition)
  } else if (common.isBetween(currentPOS[1], pitchHeight - (pitchHeight / 3), (pitchHeight - (pitchHeight / 6) + 5))) {
    if (oppositionNearPlayer(playerInformation, 10, 10)) return [30, 20, 20, 10, 0, 0, 0, 20, 0, 0, 0]
    return [70, 10, 10, 0, 0, 0, 0, 10, 0, 0, 0]
  } else if (common.isBetween(currentPOS[1], (pitchHeight / 3), (pitchHeight - (pitchHeight / 3)))) {
    if (oppositionNearPlayer(playerInformation, 10, 10)) return [0, 20, 30, 20, 0, 0, 20, 0, 0, 0, 10]
    else if (skill.shooting > 85) return [10, 10, 30, 0, 0, 0, 50, 0, 0, 0, 0]
    else if (position === 'LM' || position === 'CM' || position === 'RM') return [0, 10, 10, 10, 0, 0, 0, 30, 40, 0, 0]
    else if (position === 'ST') return [0, 0, 0, 0, 0, 0, 0, 50, 50, 0, 0]
    return [0, 0, 10, 0, 0, 0, 0, 60, 20, 0, 10]
  } else if (oppositionNearPlayer(playerInformation, 10, 10)) return [0, 0, 0, 0, 0, 0, 0, 10, 0, 70, 20]
  else if (position === 'LM' || position === 'CM' || position === 'RM') return [0, 0, 30, 0, 0, 0, 0, 30, 40, 0, 0]
  else if (position === 'ST') return [0, 0, 0, 0, 0, 0, 0, 50, 50, 0, 0]
  return [0, 0, 40, 0, 0, 0, 0, 30, 0, 20, 10]
}

function topTeamPlayerHasBallInBottomPenaltyBox(matchDetails, player, team, opposition) {
  let playerInformation = setPositions.closestPlayerToPosition(player, opposition, player.currentPOS)
  let ownPlayerInformation = setPositions.closestPlayerToPosition(player, team, player.currentPOS)
  let tmateProximity = [Math.abs(ownPlayerInformation.proxPOS[0]), Math.abs(ownPlayerInformation.proxPOS[1])]
  let closePlayerPosition = playerInformation.thePlayer.currentPOS
  const [pitchWidth, pitchHeight] = matchDetails.pitchSize
  let {
    currentPOS, skill
  } = player
  let halfRange = pitchHeight - (skill.shooting / 2)
  let shotRange = pitchHeight - skill.shooting
  if (checkPositionInBottomPenaltyBoxClose(currentPOS, pitchWidth, pitchHeight)) {
    if (oppositionNearPlayer(playerInformation, 6, 6)) {
      if (checkOppositionBelow(closePlayerPosition, currentPOS)) {
        if (checkTeamMateSpaceClose(tmateProximity, -10, 10, -10, 10)) return [20, 0, 70, 0, 0, 0, 0, 10, 0, 0, 0]
        else if (common.isBetween(currentPOS[1], halfRange, pitchHeight)) return [100, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
        else if (common.isBetween(currentPOS[1], shotRange, pitchHeight)) return [70, 0, 0, 0, 0, 0, 0, 30, 0, 0, 0]
        return [20, 0, 0, 0, 0, 0, 0, 40, 20, 0, 0]
      } else if (checkTeamMateSpaceClose(tmateProximity, -10, 10, -4, 10)) {
        if (common.isBetween(currentPOS[1], halfRange, pitchHeight)) return [90, 0, 10, 0, 0, 0, 0, 0, 0, 0, 0]
        else if (common.isBetween(currentPOS[1], shotRange, pitchHeight)) return [50, 0, 20, 0, 0, 0, 0, 30, 0, 0, 0]
        return [20, 0, 30, 0, 0, 0, 0, 30, 20, 0, 0]
      } else if (common.isBetween(currentPOS[1], halfRange, pitchHeight)) return [90, 0, 10, 0, 0, 0, 0, 0, 0, 0, 0]
      else if (common.isBetween(currentPOS[1], shotRange, pitchHeight)) return [70, 0, 0, 0, 0, 0, 0, 30, 0, 0, 0]
      return [20, 0, 0, 0, 0, 0, 0, 50, 30, 0, 0]
    } else if (checkTeamMateSpaceClose(tmateProximity, -10, 10, -4, 10)) {
      if (common.isBetween(currentPOS[1], halfRange, pitchHeight)) return [90, 0, 10, 0, 0, 0, 0, 0, 0, 0, 0]
      else if (common.isBetween(currentPOS[1], shotRange, pitchHeight)) return [50, 0, 20, 0, 0, 0, 0, 30, 0, 0, 0]
      return [20, 0, 30, 0, 0, 0, 0, 30, 20, 0, 0]
    } else if (common.isBetween(currentPOS[1], halfRange, pitchHeight)) return [100, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    else if (common.isBetween(currentPOS[1], shotRange, pitchHeight)) return [60, 0, 0, 0, 0, 0, 0, 40, 0, 0, 0]
    return [30, 0, 0, 0, 0, 0, 0, 40, 30, 0, 0]
  } else if (common.isBetween(currentPOS[1], shotRange, pitchHeight)) return [50, 0, 20, 0, 0, 0, 0, 30, 0, 0, 0]
  else if (oppositionNearPlayer(playerInformation, 6, 6)) return [10, 0, 70, 0, 0, 0, 0, 20, 0, 0, 0]
  return [70, 0, 20, 0, 0, 0, 0, 10, 0, 0, 0]
}

function bottomTeamPlayerHasBall(matchDetails, player, team, opposition) {
  let playerInformation = setPositions.closestPlayerToPosition(player, opposition, player.currentPOS)
  const [pitchWidth, pitchHeight] = matchDetails.pitchSize
  let {
    position, currentPOS, skill
  } = player
  if (position === 'GK' && oppositionNearPlayer(playerInformation, 10, 25)) return [0, 0, 10, 0, 0, 0, 0, 10, 0, 40, 40]
  else if (position === 'GK') return [0, 0, 50, 0, 0, 0, 0, 10, 0, 20, 20]
  else if (onTopCornerBoundary(currentPOS, pitchWidth)) return [0, 0, 20, 80, 0, 0, 0, 0, 0, 0, 0]
  else if (checkPositionInTopPenaltyBox(currentPOS, pitchWidth, pitchHeight)) {
    return bottomTeamPlayerHasBallInTopPenaltyBox(matchDetails, player, team, opposition)
  } else if (common.isBetween(currentPOS[1], (pitchHeight / 6) - 5, pitchHeight / 3)) {
    if (oppositionNearPlayer(playerInformation, 10, 10)) return [30, 20, 20, 10, 0, 0, 0, 20, 0, 0, 0]
    return [70, 10, 10, 0, 0, 0, 0, 10, 0, 0, 0]
  } else if (common.isBetween(currentPOS[1], (pitchHeight / 3), (2 * (pitchHeight / 3)))) {
    return bottomTeamPlayerHasBallInMiddle(playerInformation, position, skill)
  } else if (oppositionNearPlayer(playerInformation, 10, 10)) return [0, 0, 0, 0, 0, 0, 0, 10, 0, 70, 20]
  else if (position === 'LM' || position === 'CM' || position === 'RM') return [0, 0, 30, 0, 0, 0, 0, 30, 40, 0, 0]
  else if (position === 'ST') return [0, 0, 0, 0, 0, 0, 0, 50, 50, 0, 0]
  return [0, 0, 30, 0, 0, 0, 0, 50, 0, 10, 10]
}

function bottomTeamPlayerHasBallInMiddle(playerInformation, position, skill) {
  if (oppositionNearPlayer(playerInformation, 10, 10)) return [0, 20, 30, 20, 0, 0, 0, 20, 0, 0, 10]
  else if (skill.shooting > 85) return [10, 10, 30, 0, 0, 0, 0, 50, 0, 0, 0]
  else if (position === 'LM' || position === 'CM' || position === 'RM') return [0, 10, 10, 10, 0, 0, 0, 30, 40, 0, 0]
  else if (position === 'ST') return [0, 0, 0, 0, 0, 0, 0, 50, 50, 0, 0]
  return [0, 0, 10, 0, 0, 0, 0, 60, 20, 0, 10]
}

function bottomTeamPlayerHasBallInTopPenaltyBox(matchDetails, player, team, opposition) {
  let playerInformation = setPositions.closestPlayerToPosition(player, opposition, player.currentPOS)
  let ownPlayerInformation = setPositions.closestPlayerToPosition(player, team, player.currentPOS)
  let tmateProximity = [Math.abs(ownPlayerInformation.proxPOS[0]), Math.abs(ownPlayerInformation.proxPOS[1])]
  let closePlayerPosition = playerInformation.thePlayer.currentPOS
  const [pitchWidth, pitchHeight] = matchDetails.pitchSize
  let {
    currentPOS, skill
  } = player
  if (checkPositionInTopPenaltyBoxClose(currentPOS, pitchWidth, pitchHeight)) {
    if (oppositionNearPlayer(playerInformation, 20, 20)) {
      if (checkOppositionAhead(closePlayerPosition, currentPOS)) {
        if (checkTeamMateSpaceClose(tmateProximity, -10, 10, -10, 10)) return [20, 0, 70, 0, 0, 0, 0, 10, 0, 0, 0]
        else if (common.isBetween(currentPOS[1], 0, (skill.shooting / 2))) return [100, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
        else if (common.isBetween(currentPOS[1], 0, skill.shooting)) return [70, 0, 0, 0, 0, 0, 0, 30, 0, 0, 0]
        return [20, 0, 0, 0, 0, 0, 0, 40, 20, 0, 0]
      } else if (checkTeamMateSpaceClose(tmateProximity, -10, 10, -4, 10)) {
        if (common.isBetween(currentPOS[1], 0, (skill.shooting / 2))) return [90, 0, 10, 0, 0, 0, 0, 0, 0, 0, 0]
        else if (common.isBetween(currentPOS[1], 0, skill.shooting)) return [50, 0, 20, 0, 0, 0, 0, 30, 0, 0, 0]
        return [20, 0, 30, 0, 0, 0, 0, 30, 20, 0, 0]
      } else if (common.isBetween(currentPOS[1], 0, (skill.shooting / 2))) return [90, 0, 10, 0, 0, 0, 0, 0, 0, 0, 0]
      else if (common.isBetween(currentPOS[1], 0, skill.shooting)) return [70, 0, 0, 0, 0, 0, 0, 30, 0, 0, 0]
      return [20, 0, 0, 0, 0, 0, 0, 50, 30, 0, 0]
    } else if (checkTeamMateSpaceClose(tmateProximity, -10, 10, -4, 10)) {
      if (common.isBetween(currentPOS[1], 0, (skill.shooting / 2))) return [90, 0, 10, 0, 0, 0, 0, 0, 0, 0, 0]
      else if (common.isBetween(currentPOS[1], 0, skill.shooting)) return [50, 0, 20, 0, 0, 0, 0, 30, 0, 0, 0]
      return [20, 0, 30, 0, 0, 0, 0, 30, 20, 0, 0]
    } else if (common.isBetween(currentPOS[1], 0, (skill.shooting / 2))) return [100, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    else if (common.isBetween(currentPOS[1], 0, skill.shooting)) return [60, 0, 0, 0, 0, 0, 0, 40, 0, 0, 0]
    return [30, 0, 0, 0, 0, 0, 0, 40, 30, 0, 0]
  } else if (common.isBetween(currentPOS[1], 0, skill.shooting)) return [50, 0, 20, 0, 0, 0, 0, 30, 0, 0, 0]
  else if (checkOppositionAhead(closePlayerPosition, currentPOS)) return [20, 0, 0, 0, 0, 0, 0, 80, 0, 0, 0]
  return [50, 0, 20, 20, 0, 0, 0, 10, 0, 0, 0]
}

function oppositionNearPlayer(oppositionPlayer, spaceX, spaceY) {
  let oppositionProximity = [Math.abs(oppositionPlayer.proxPOS[0]), Math.abs(oppositionPlayer.proxPOS[1])]
  if (oppositionProximity[0] < spaceX && oppositionProximity[1] < spaceY) return true
  return false
}

function checkTeamMateSpaceClose(tmateProximity, lowX, highX, lowY, highY) {
  if (common.isBetween(tmateProximity[0], lowX, highX) && common.isBetween(tmateProximity[1], lowY, highY)) return true
  return false
}

function checkOppositionAhead(closePlayerPosition, currentPOS) {
  let closePlyX = common.isBetween(closePlayerPosition[0], currentPOS[0] - 4, currentPOS[0] + 4)
  if (closePlyX && closePlayerPosition[1] < currentPOS[1]) return true
  return false
}

function checkOppositionBelow(closePlayerPosition, currentPOS) {
  let closePlyX = common.isBetween(closePlayerPosition[0], currentPOS[0] - 4, currentPOS[0] + 4)
  if (closePlyX && closePlayerPosition[1] > currentPOS[1]) return true
  return false
}

function playerDoesNotHaveBall(player, ballX, ballY, matchDetails) {
  const [pitchWidth, pitchHeight] = matchDetails.pitchSize
  let {
    position, currentPOS, originPOS
  } = player
  if (position === 'GK') return [0, 0, 0, 0, 0, 0, 0, 60, 40, 0, 0]
  else if (common.isBetween(ballX, -20, 20) && common.isBetween(ballY, -20, 20)) {
    return noBallNotGK2CloseBall(matchDetails, currentPOS, originPOS, pitchWidth, pitchHeight)
  } else if (common.isBetween(ballX, -40, 40) && common.isBetween(ballY, -40, 40)) {
    return noBallNotGK4CloseBall(matchDetails, currentPOS, originPOS, pitchWidth, pitchHeight)
  } else if (common.isBetween(ballX, -80, 80) && common.isBetween(ballY, -80, 80)) {
    if (matchDetails.ball.withPlayer === false) return [0, 0, 0, 0, 0, 0, 0, 60, 40, 0, 0]
    return [0, 0, 0, 0, 0, 40, 0, 30, 30, 0, 0]
  }
  return [0, 0, 0, 0, 0, 10, 0, 50, 30, 0, 0]
}

function noBallNotGK4CloseBall(matchDetails, currentPOS, originPOS, pitchWidth, pitchHeight) {
  if (originPOS[1] > (pitchHeight / 2)) {
    return noBallNotGK4CloseBallBottomTeam(matchDetails, currentPOS, pitchWidth, pitchHeight)
  }
  if (checkPositionInTopPenaltyBox(currentPOS, pitchWidth, pitchHeight)) {
    if (matchDetails.ball.withPlayer === false) return [0, 0, 0, 0, 0, 0, 0, 20, 80, 0, 0]
    return [0, 0, 0, 0, 40, 0, 20, 10, 30, 0, 0]
  } else if (matchDetails.ball.withPlayer === false) return [0, 0, 0, 0, 0, 0, 0, 20, 80, 0, 0]
  return [0, 0, 0, 0, 50, 0, 50, 0, 0, 0, 0]
}

function noBallNotGK4CloseBallBottomTeam(matchDetails, currentPOS, pitchWidth, pitchHeight) {
  if (checkPositionInBottomPenaltyBox(currentPOS, pitchWidth, pitchHeight)) {
    if (matchDetails.ball.withPlayer === false) return [0, 0, 0, 0, 0, 0, 0, 20, 80, 0, 0]
    return [0, 0, 0, 0, 40, 0, 20, 10, 30, 0, 0]
  } else if (matchDetails.ball.withPlayer === false) return [0, 0, 0, 0, 0, 0, 0, 20, 80, 0, 0]
  return [0, 0, 0, 0, 50, 0, 50, 0, 0, 0, 0]
}

function noBallNotGK2CloseBall(matchDetails, currentPOS, originPOS, pitchWidth, pitchHeight) {
  if (originPOS[1] > (pitchHeight / 2)) {
    return noBallNotGK2CloseBallBottomTeam(matchDetails, currentPOS, pitchWidth, pitchHeight)
  }
  if (checkPositionInTopPenaltyBox(currentPOS, pitchWidth, pitchHeight)) {
    if (matchDetails.ball.withPlayer === false) return [0, 0, 0, 0, 0, 0, 0, 20, 80, 0, 0]
    return [0, 0, 0, 0, 40, 0, 20, 10, 30, 0, 0]
  } else if (matchDetails.ball.withPlayer === false) return [0, 0, 0, 0, 0, 0, 0, 20, 80, 0, 0]
  return [0, 0, 0, 0, 70, 10, 20, 0, 0, 0, 0]
}

function noBallNotGK2CloseBallBottomTeam(matchDetails, currentPOS, pitchWidth, pitchHeight) {
  if (checkPositionInBottomPenaltyBox(currentPOS, pitchWidth, pitchHeight)) {
    if (matchDetails.ball.withPlayer === false) return [0, 0, 0, 0, 0, 0, 0, 20, 80, 0, 0]
    return [0, 0, 0, 0, 50, 0, 10, 20, 20, 0, 0]
  }
  if (matchDetails.ball.withPlayer === false) return [0, 0, 0, 0, 0, 0, 0, 20, 80, 0, 0]
  return [0, 0, 0, 0, 70, 10, 20, 0, 0, 0, 0]
}

function checkPositionInBottomPenaltyBox(position, pitchWidth, pitchHeight) {
  let yPos = common.isBetween(position[0], (pitchWidth / 4) - 5, pitchWidth - (pitchWidth / 4) + 5)
  let xPos = common.isBetween(position[1], pitchHeight - (pitchHeight / 6) + 5, pitchHeight)
  if (yPos && xPos) return true
  return false
}

function checkPositionInBottomPenaltyBoxClose(position, pitchWidth, pitchHeight) {
  let yPos = common.isBetween(position[0], (pitchWidth / 3) - 5, pitchWidth - (pitchWidth / 3) + 5)
  let xPos = common.isBetween(position[1], (pitchHeight - (pitchHeight / 12) + 5), pitchHeight)
  if (yPos && xPos) return true
  return false
}

function checkPositionInTopPenaltyBox(position, pitchWidth, pitchHeight) {
  let xPos = common.isBetween(position[0], (pitchWidth / 4) - 5, pitchWidth - (pitchWidth / 4) + 5)
  let yPos = common.isBetween(position[1], 0, (pitchHeight / 6) - 5)
  if (yPos && xPos) return true
  return false
}

function checkPositionInTopPenaltyBoxClose(position, pitchWidth, pitchHeight) {
  let xPos = common.isBetween(position[0], (pitchWidth / 3) - 5, pitchWidth - (pitchWidth / 3) + 5)
  let yPos = common.isBetween(position[1], 0, (pitchHeight / 12) - 5)
  if (yPos && xPos) return true
  return false
}

function onBottomCornerBoundary(position, pitchWidth, pitchHeight) {
  if (position[1] == pitchHeight && (position[0] == 0 || position[0] == pitchWidth)) return true
  return false
}

function onTopCornerBoundary(position, pitchWidth) {
  if (position[1] == 0 && (position[0] == 0 || position[0] == pitchWidth)) return true
  return false
}

function populatePossibleActions(possibleActions, a, b, c, d, e, f, g, h, i, j, k) {
  //a-shoot, b-throughBall, c-pass, d-cross, e-tackle, f-intercept
  //g-slide, h-run, i-sprint j-cleared k-boot
  possibleActions[0].points = a
  possibleActions[1].points = b
  possibleActions[2].points = c
  possibleActions[3].points = d
  possibleActions[4].points = e
  possibleActions[5].points = f
  possibleActions[6].points = g
  possibleActions[7].points = h
  possibleActions[8].points = i
  possibleActions[9].points = j
  possibleActions[10].points = k
  return possibleActions
}

function populateActionsJSON() {
  return [{
    'name': 'shoot',
    'points': 0
  }, {
    'name': 'throughBall',
    'points': 0
  }, {
    'name': 'pass',
    'points': 0
  }, {
    'name': 'cross',
    'points': 0
  }, {
    'name': 'tackle',
    'points': 0
  }, {
    'name': 'intercept',
    'points': 0
  }, {
    'name': 'slide',
    'points': 0
  }, {
    'name': 'run',
    'points': 0
  }, {
    'name': 'sprint',
    'points': 0
  }, {
    'name': 'cleared',
    'points': 0
  }, {
    'name': 'boot',
    'points': 0
  }]
}

function resolveTackle(player, team, opposition, matchDetails) {
  matchDetails.iterationLog.push(`Tackle attempted by: ${player.name}`)
  let tackleDetails = {
    'injuryHigh': 1500,
    'injuryLow': 1400,
    'increment': 1
  }
  let index = opposition.players.findIndex(function(thisPlayer) {
    return thisPlayer.playerID === matchDetails.ball.Player
  })
  let thatPlayer
  if (index) thatPlayer = opposition.players[index]
  else return false
  player.stats.tackles.total++
  if (wasFoul(10, 18)) {
    setFoul(matchDetails, team, player, thatPlayer)
    return true
  }
  if (calcTackleScore(player.skill.tackling, 5) > calcRetentionScore(thatPlayer.skill.tackling, 5)) {
    setSuccessTackle(matchDetails, team, opposition, player, thatPlayer, tackleDetails)
    return false
  }
  setFailedTackle(matchDetails, player, thatPlayer, tackleDetails)
  return false
}

function resolveSlide(player, team, opposition, matchDetails) {
  matchDetails.iterationLog.push(`Slide tackle attempted by: ${player.name}`)
  let tackleDetails = {
    'injuryHigh': 1500,
    'injuryLow': 1400,
    'increment': 3
  }
  let index = opposition.players.findIndex(function(thisPlayer) {
    return thisPlayer.playerID === matchDetails.ball.Player
  })
  let thatPlayer
  if (index) thatPlayer = opposition.players[index]
  else return false
  player.stats.tackles.total++
  if (wasFoul(11, 20)) {
    setFoul(matchDetails, team, player, thatPlayer)
    return true
  }
  if (calcTackleScore(player.skill.tackling, 5) > calcRetentionScore(thatPlayer.skill.tackling, 5)) {
    setSuccessTackle(matchDetails, team, opposition, player, thatPlayer, tackleDetails)
    return false
  }
  setFailedTackle(matchDetails, player, thatPlayer, tackleDetails)
  return false
}

function setFailedTackle(matchDetails, player, thatPlayer, tackleDetails) {
  matchDetails.iterationLog.push(`Failed tackle by: ${player.name}`)
  player.stats.tackles.off++
  setInjury(matchDetails, player, thatPlayer, tackleDetails.injuryHigh, tackleDetails.injuryLow)
  setPostTacklePosition(matchDetails, thatPlayer, player, tackleDetails.increment)
}

function setSuccessTackle(matchDetails, team, opposition, player, thatPlayer, tackleDetails) {
  setPostTackleBall(matchDetails, team, opposition, player)
  matchDetails.iterationLog.push(`Successful tackle by: ${player.name}`)
  player.stats.tackles.on++
  setInjury(matchDetails, thatPlayer, player, tackleDetails.injuryLow, tackleDetails.injuryHigh)
  setPostTacklePosition(matchDetails, player, thatPlayer, tackleDetails.increment)
}

function calcTackleScore(skill, diff) {
  return ((parseInt(skill.tackling, 10) + parseInt(skill.strength, 10)) / 2) + common.getRandomNumber(-diff, diff)
}

function calcRetentionScore(skill, diff) {
  return ((parseInt(skill.agility, 10) + parseInt(skill.strength, 10)) / 2) + common.getRandomNumber(-diff, diff)
}

function setPostTackleBall(matchDetails, team, opposition, player) {
  player.hasBall = true
  matchDetails.ball.lastTouch.playerName = player.name
  matchDetails.ball.lastTouch.playerID = player.playerID
  matchDetails.ball.lastTouch.teamID = team.teamID
  let tempArray = player.currentPOS
  matchDetails.ball.position = tempArray.map(x => x)
  matchDetails.ball.Player = player.playerID
  matchDetails.ball.withPlayer = true
  matchDetails.ball.withTeam = team.teamID
  team.intent = 'attack'
  opposition.intent = 'defend'
}

function setPostTacklePosition(matchDetails, winningPlyr, losePlayer, increment) {
  const [, pitchHeight] = matchDetails.pitchSize
  if (losePlayer.originPOS[1] > pitchHeight / 2) {
    losePlayer.currentPOS[1] = common.upToMin(losePlayer.currentPOS[1] - increment, 0)
    matchDetails.ball.position[1] = common.upToMin(matchDetails.ball.position[1] - increment, 0)
    winningPlyr.currentPOS[1] = common.upToMax(winningPlyr.currentPOS[1] + increment, pitchHeight)
  } else {
    losePlayer.currentPOS[1] = common.upToMax(losePlayer.currentPOS[1] + increment, pitchHeight)
    matchDetails.ball.position[1] = common.upToMax(matchDetails.ball.position[1] + increment, pitchHeight)
    winningPlyr.currentPOS[1] = common.upToMin(winningPlyr.currentPOS[1] - increment, 0)
  }
}

function setInjury(matchDetails, thatPlayer, player, tackledInjury, tacklerInjury) {
  if (common.isInjured(tackledInjury)) {
    thatPlayer.injured = true
    matchDetails.iterationLog.push(`Player Injured - ${thatPlayer.name}`)
  }
  if (common.isInjured(tacklerInjury)) {
    player.injured = true
    matchDetails.iterationLog.push(`Player Injured - ${player.name}`)
  }
}

function setFoul(matchDetails, team, player, thatPlayer) {
  matchDetails.iterationLog.push(`Foul against: ${thatPlayer.name}`)
  player.stats.tackles.fouls++
  if (team.teamID === matchDetails.kickOffTeam.teamID) matchDetails.kickOffTeamStatistics.fouls++
  else matchDetails.secondTeamStatistics.fouls++
}

function wasFoul(x, y) {
  let foul = common.getRandomNumber(0, x)
  if (common.isBetween(foul, 0, (y / 2) - 1)) return true
  return false
}

function foulIntensity() {
  return common.getRandomNumber(1, 99)
}

module.exports = {
  selectAction,
  findPossActions,
  playerDoesNotHaveBall,
  topTeamPlayerHasBall,
  topTeamPlayerHasBallInBottomPenaltyBox,
  bottomTeamPlayerHasBall,
  bottomTeamPlayerHasBallInMiddle,
  bottomTeamPlayerHasBallInTopPenaltyBox,
  noBallNotGK2CloseBall,
  noBallNotGK2CloseBallBottomTeam,
  noBallNotGK4CloseBall,
  noBallNotGK4CloseBallBottomTeam,
  oppositionNearPlayer,
  checkTeamMateSpaceClose,
  checkOppositionAhead,
  checkOppositionBelow,
  checkPositionInTopPenaltyBox,
  checkPositionInTopPenaltyBoxClose,
  onBottomCornerBoundary,
  onTopCornerBoundary,
  checkPositionInBottomPenaltyBox,
  checkPositionInBottomPenaltyBoxClose,
  populatePossibleActions,
  resolveTackle,
  resolveSlide,
  calcTackleScore,
  calcRetentionScore,
  setPostTackleBall,
  setPostTacklePosition,
  setFoul,
  setInjury,
  wasFoul,
  foulIntensity
}
